lib/artsy-eventservice/artsy/event_service/rabbitmq_connection.rb

```diff
@@ -3,44 +3,56 @@
 
 module Artsy
   module EventService
-    module RabbitMQConnection
-      # get a new RabbitMQ Client
-      def create_conn
-        Bunny.new(rabbitmq_url, **bunny_params)
-      end
+    class RabbitMQConnection
+      @connection = nil
+      @mutex = Mutex.new
 
-      # Connect, do something and close the connection
-      def connect_to_rabbit
-        conn = create_conn
+      # Build a new connection to RabbitMQ
+      def self.build_connection
+        conn = Bunny.new(self.rabbitmq_url, **self.bunny_params)
         conn.start
-        yield(conn)
+        conn
+      end
+
+      # Synchronized access to the connection - rebuild if closed
+      def self.get_connection
+        @mutex.synchronize do
+          @connection = nil if @connection && @connection.closed?
+          @connection ||= self.build_connection
+        end
+      end
+
+      # Get a channel from the connection - synchronized access to create_channel is provided by Bunny
+      def self.get_channel
+        channel = self.get_connection.create_channel
```

Comment:
Wondering if there's a race condition here, as `get_connection` and `create_channel` use their own individual internal Mutexes for synchronization... i.e. one thread could call `get_connection` and before it calls `create_channel`, a new thread calls `get_connection` but determines `@connection` to be closed and sets it to `nil`.  Then the first thread will get a `Undefined method create_channel for nil object`...

Reply:
But in this case, calling `create_channel` on a closed connection would also fail... wondering if there's a better way to handle it without wrapping one Mutex inside the other...

Reply:
@joeyAghion @yuki24 as a side note, regarding our earlier discussion (link?), I would not move the `EventService.post_event` calls from Sidekiq into Puma workers, until we ensure this can recover from these kinds of [ConnectionLevelExceptions](http://reference.rubybunny.info/Bunny/ConnectionLevelException.html) without blocking the request... possibly by making each call to `EventService.post_event` in a separate thread...

Reply:
This looks fine to me, but I'll write a script to make sure it won't fail due to a NoMethodError.

Reply:
You are right that this could fail when calling `create_channel`: https://gist.github.com/yuki24/cce8c32b757db18177d37dd46835551c

I agree we should send events through sidekiq for now.

Reply:
Thanks for the gist yuki!  I'm currently trying to port it into an integration test :)

Reply:
Since `automatically_recover` and `recover_from_connection_close` are both set to true and `recovery_attempts` is set to infinite (see [Bunny::Session#initialize](http://reference.rubybunny.info/Bunny/Session.html#initialize-instance_method)), why do we need to manually set the connection to `nil` ourselves ever? 

In the event of a network issue we can then just expect the Session to raise an error of some sort as it tries to recover by itself - however, it's not clear from the doc what this is. In any case, it doesn't seem to me that creating a new session object will fix any networking issues that the client can't recover from itself.

Reply:
@cavvia yes this is why I updated Bunny to 2.7.1 - I observed some issues when testing publishing concurrently across a thread pool while randomly killing the connection in another, and saw some `Bunny::ChannelAlreadyClosed` exceptions it could not recover from, but the update seems to have fixed this